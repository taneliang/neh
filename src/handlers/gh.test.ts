import fetchMock from 'jest-fetch-mock';
import handler from './gh';

describe('gh handler', () => {
  describe('p handler', () => {
    test('should redirect if at least one user is returned', async () => {
      fetchMock.enableMocks();

      const htmlUrl = 'https://github.com/taneliang';
      // eslint-disable-next-line @typescript-eslint/camelcase
      const expectedJsonResponse = { items: [{ html_url: htmlUrl }] };
      fetchMock.mockOnce(async () => JSON.stringify(expectedJsonResponse));

      const response = await handler.handle(['p', 'some query']);
      expect(response.status).toBe(302);
      expect(response.headers.get('location')).toEqual(htmlUrl);

      fetchMock.disableMocks();
    });

    test('should return error message if GitHub returns empty items array', async () => {
      fetchMock.enableMocks();

      const expectedJsonResponse = { items: [] };
      fetchMock.mockOnce(async () => JSON.stringify(expectedJsonResponse));

      const response = await handler.handle(['p', 'some query']);
      expect(response).toMatchSnapshot();

      fetchMock.disableMocks();
    });

    test('should return error message if GitHub API returns error', async () => {
      fetchMock.enableMocks();

      fetchMock.mockOnce(async () => ({
        init: {
          status: 403,
          statusText: 'Forbidden',
        },
        body: 'Some message',
      }));

      const response = await handler.handle(['p', 'some query']);
      expect(response).toMatchSnapshot();

      fetchMock.disableMocks();
    });
  });

  describe('r handler', () => {
    test('should redirect if at least one repository is returned', async () => {
      fetchMock.enableMocks();

      const htmlUrl = 'https://github.com/taneliang/neh';
      // eslint-disable-next-line @typescript-eslint/camelcase
      const expectedJsonResponse = { items: [{ html_url: htmlUrl }] };
      fetchMock.mockOnce(async () => JSON.stringify(expectedJsonResponse));

      const response = await handler.handle(['r', 'some query']);
      expect(response.status).toBe(302);
      expect(response.headers.get('location')).toEqual(htmlUrl);

      fetchMock.disableMocks();
    });

    test('should return error message if GitHub returns empty items array', async () => {
      fetchMock.enableMocks();

      const expectedJsonResponse = { items: [] };
      fetchMock.mockOnce(async () => JSON.stringify(expectedJsonResponse));

      const response = await handler.handle(['r', 'some query']);
      expect(response).toMatchSnapshot();

      fetchMock.disableMocks();
    });

    test('should return error message if GitHub API returns error', async () => {
      fetchMock.enableMocks();

      fetchMock.mockOnce(async () => ({
        init: {
          status: 403,
          statusText: 'Forbidden',
        },
        body: 'Some message',
      }));

      const response = await handler.handle(['r', 'some query']);
      expect(response).toMatchSnapshot();

      fetchMock.disableMocks();
    });
  });
});
