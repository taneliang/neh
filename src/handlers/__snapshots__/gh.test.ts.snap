// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gh handler p handler should return error message if GitHub API returns error 1`] = `
Response {
  "body": Blob {
    "parts": Array [
      "GitHub API returned 403 Forbidden. Some message",
    ],
    "type": "",
  },
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {},
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 500,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;

exports[`gh handler p handler should return error message if GitHub returns empty items array 1`] = `
Response {
  "body": Blob {
    "parts": Array [
      "No users found for query some query.",
    ],
    "type": "",
  },
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {},
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 500,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;

exports[`gh handler r handler should return error message if GitHub API returns error 1`] = `
Response {
  "body": Blob {
    "parts": Array [
      "GitHub API returned 403 Forbidden. Some message",
    ],
    "type": "",
  },
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {},
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 500,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;

exports[`gh handler r handler should return error message if GitHub returns empty items array 1`] = `
Response {
  "body": Blob {
    "parts": Array [
      "No repositories found for query some query.",
    ],
    "type": "",
  },
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {},
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 500,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;
